lib/cn/cn_tree.c

```diff
@@ -1485,98 +1485,64 @@ cn_comp_update_spill(struct cn_compaction_work *work, struct kvset **kvsets)
 /**
  * cn_comp_update_split() - update tree after a node split operation
  */
-static merr_t
+static void
 cn_comp_update_split(
     struct cn_compaction_work *w,
     struct kvset *const       *kvsets,
-    const uint64_t             nodeidv[static 2])
+    struct cn_tree_node       *nodev[static 2])
 {
     struct cn_tree *tree = w->cw_tree;
     struct kvset_list_entry *le, *tmp;
     struct list_head retired_kvsets;
-    struct cn_tree_node *left = NULL, *right = w->cw_node;
-    char rekey[HSE_KVS_KEY_LEN_MAX];
-    uint k = 0, reklen;
-    merr_t err = 0;
+    struct cn_tree_node *left = nodev[LEFT], *right = nodev[RIGHT];
+    uint k = 0;
 
-    if (ev(w->cw_err))
-        return w->cw_err;
+    if (w->cw_err)
+        return;
 
     INIT_LIST_HEAD(&retired_kvsets);
 
-    /* Allocate a new left node and add the split output kvsets on the left to this node.
-     * This need not be done under the tree lock as this new node is not published yet.
+    /* Add the left half of the split kvsets to the 'left' node.
+     * This need not be done under the tree lock as the new left node is not published yet.
      */
-    if (nodeidv[0] != CN_TREE_INVALID_NODEID) {
-        left = cn_node_alloc(tree, nodeidv[0]);
-        if (!left)
-            return merr(ENOMEM);
-
+    if (left) {
+        assert(list_empty(&left->tn_kvset_list));
         for (k = 0; k < w->cw_kvset_cnt; k++) {
             if (kvsets[k])
                 kvset_list_add_tail(kvsets[k], &left->tn_kvset_list);
         }
 
-        w->cw_split.nodev[0] = left;
-    }
-
-    if (nodeidv[1] != CN_TREE_INVALID_NODEID) {
-        /* The 'right' node is protected by an exclusive compaction token, so the max key cannot
-         * change while a node split is in progress
-         */
-        cn_tree_node_get_max_key(right, rekey, sizeof(rekey), &reklen);
+        w->cw_split.nodev[LEFT] = left;
     }
 
     rmlock_wlock(&tree->ct_lock);
-
-    do {
+    {
         /* Move all the source kvsets from the 'right' node to the retired list.
          */
         list_splice(&right->tn_kvset_list, &retired_kvsets);
         INIT_LIST_HEAD(&right->tn_kvset_list);
 
         /* Add the right half of the split kvsets to the 'right' node.
          */
-        if (nodeidv[1] != CN_TREE_INVALID_NODEID) {
-            right->tn_nodeid = nodeidv[1];
-
+        if (right) {
```

Comment:
When would you not have a right node?

Reply:
This should not happen under any normal circumstances.

This check exists just to protect against a pathological case where a node split moves all the kvsets from the source node to one side either left or right.